### ISSUE TITLE
[Bug] DataLoader worker segmentation fault in multi-task training with long sequences and numerical instability

### ISSUE THREAD

**Issue by inyohh** [pytorch/pytorch#181853]
### 🐛 Describe the bug

GitHub Issue Title: 
  [Bug] DataLoader worker segmentation fault in multi-task training with long sequences and numerical instability

  Description:
  A DataLoader worker process encountered a segmentation fault during a multi-task BERT training loop. The crash occurred after several thousand steps, consistently preceded by
  NaN/Inf warnings in the input tensors.

  Error Message:

   1 (train_bohb pid=420171, ip=192.168.235.225) ERROR: Unexpected segmentation fault encountered in worker.
   2 (train_bohb pid=420171, ip=192.168.235.225) ERROR: Training failed: DataLoader worker (pid(s) 423100) exited unexpectedly

  Environment:
   - OS: Linux (Ubuntu 22.04/24.04 candidate)
   - GPU: NVIDIA GeForce RTX 4090 (24GB VRAM)
   - Driver Version: 580.126.09
   - CUDA Version: 13.0
   - Python Version: 3.12
   - PyTorch/Transformers: (Please fill in the versions from your specific env, e.g., PyTorch 2.4.0)

  Context & Observed Behavior:
   - Training Config: max_length=506, num_workers=4, per_device_eval_batch_size=64.
   - Symptoms: 
       1. Training runs stably for ~3 hours (up to epoch 3.45).
       2. Numerical instability occurs: WARNING:tensorboardX.x2num:NaN or Inf found in input tensor.
       3. Shortly after, the DataLoader worker triggers a segmentation fault.
   - Hypothesis: The combination of near-limit sequence lengths (506/512) and NaN gradients might be triggering an illegal memory access in

### Versions

Context:
   - Batch Size: 32 (Train) / 64 (Eval)
   - Max Sequence Length: 506 (near the 512 BERT limit)
   - Num Workers: 4
   - Shared Memory (/dev/shm): 32GB (Adequate)

  Preliminary Analysis:
  The crash seems to be triggered when the model encounters numerical instability (NaN/Inf in tensors) combined with high memory/VRAM pressure from long sequences. It does not appear
  to be a simple OOM, as the system does not trigger the OOM Killer, but rather a segmentation fault in the worker process, possibly within a C++ extension or a CUDA kernel invoked
  during data collation/processing.

  Steps to Reproduce (if applicable):
   1. Use a Multi-task BERT model with max_length set to 500+.
   2. Train with num_workers > 0 and fp16=True.
   3. Introduce or wait for a condition where gradients/loss become inf.

cc @andrewkho @divyanshk @SsnL @VitalyFedyunin @dzhulgakov @scotts

**Comment by drisspg** [comment]
Can you add some more concrete steps to reproduce

**Comment by inyohh** [comment]
Reproduction Guide: Intermittent DataLoader Worker Segfault on Intel 13th/14th Gen

  1. Environment Setup (Critical)
  The issue appears to be specific to high-end Intel CPUs (e.g., i9-13900K) where Split Lock Mitigation is active.

   * Check/Enable Mitigation:

   1     # Ensure split lock mitigation is enabled (this is the default on many modern kernels)
   2     sudo sysctl -w kernel.split_lock_mitigate=1
   * Platform: Linux (e.g., Ubuntu 24.04+), Python 3.12, PyTorch 2.x, CUDA 13.0.

  2. Minimal Reproduction Script
  This script simulates a multi-task BERT-like data loading pattern with long sequences and intentional numerical instability.

```import torch
from torch.utils.data import Dataset, DataLoader
import numpy as np
import os


class StressDataset(Dataset):
    def __init__(self, size=100000, seq_len=506):
        self.size = size
        self.seq_len = seq_len

    def __len__(self):
        return self.size

    def __getitem__(self, idx):
        # Simulate long sequences near the 512 limit
        input_ids = torch.randint(0, 20000, (self.seq_len,))
        attention_mask = torch.ones((self.seq_len,))

        # Simulate numerical instability (intermittent Inf/NaN)
        # This sometimes triggers illegal memory access in C++ extensions/kernels
        labels = torch.tensor([float(idx % 10)])
        if idx % 500 == 0:
            labels = torch.tensor([float("inf")])

        return {
            "input_ids": input_ids,
            "attention_mask": attention_mask,
            "labels": labels,
        }


def collate_fn(batch):
    # Standard collation but with large tensors
    return {key: torch.stack([item[key] for item in batch]) for key in batch[0].keys()}


if __name__ == "__main__":
    dataset = StressDataset()
    # Using high num_workers and pin_memory to maximize shared memory usage
    dataloader = DataLoader(
        dataset, batch_size=64, num_workers=4, pin_memory=True, collate_fn=collate_fn
    )

    print(f"Starting Stress Test (PID: {os.getpid()})...")
    try:
        for epoch in range(10):
            for i, batch in enumerate(dataloader):
                if i % 100 == 0:
                    # Simulate a tiny bit of GPU work
                    _ = batch["input_ids"].to("cuda")
                    if i % 1000 == 0:
                        print(f"Epoch {epoch}, Step {i} processed successfully.")
    except Exception as e:
        print(f"Caught Python Exception: {e}")
    # Note: If it's a Segfault, Python won't catch it; the process will just die.
```

  3. Steps to Reproduce
   1. Run the script on a machine with kernel.split_lock_mitigate=1.
   2. Observe the output. The crash is intermittent (non-deterministic).
   3. Monitor system logs simultaneously in another terminal:

   1     # Look for "split_lock" or "segfault"
   2     sudo journalctl -kf | grep -iE "split lock|segfault"

  4. Expected vs. Actual Behavior
   * Expected: The DataLoader should either raise a Python ValueError (for Inf values if handled) or complete the loop.
   * Actual: The worker process terminates abruptly with ERROR: Unexpected segmentation fault encountered in worker or DataLoader worker exited unexpectedly. No Python Traceback is
     provided because the crash happens in the C++/CUDA layer or is induced by a kernel-level SIGKILL/SIGBUS due to split-lock detection.

  ---

  Why these steps? (Technical justification for the devs)
   * seq_len=506: Long sequences increase the probability of unaligned memory access in vectorized operations.
   * split_lock_mitigate=1: On modern Intel CPUs, atomic operations crossing cache lines (split locks) are penalized. If PyTorch's internal buffer management or a specific CUDA kernel
     performs an unaligned atomic op, the kernel mitigation might terminate the process or cause enough delay to trigger a race condition in the multiprocessing queue.
   * No Traceback: The lack of a traceback is a hallmark of a signal-level crash (SIGSEGV) rather than a logic-level Python exception.

**Comment by inyohh** [comment]
Reproduction Guide: Intermittent DataLoader Worker Segfault on Intel 13th/14th Gen

  1. Environment Setup (Critical)
  The issue appears to be specific to high-end Intel CPUs (e.g., i9-13900K) where Split Lock Mitigation is active.

   * Check/Enable Mitigation:

   1     # Ensure split lock mitigation is enabled (this is the default on many modern kernels)
   2     sudo sysctl -w kernel.split_lock_mitigate=1
   * Platform: Linux (e.g., Ubuntu 24.04+), Python 3.12, PyTorch 2.x, CUDA 13.0.

  2. Minimal Reproduction Script
  This script simulates a multi-task BERT-like data loading pattern with long sequences and intentional numerical instability.

```import torch
from torch.utils.data import Dataset, DataLoader
import numpy as np
import os


class StressDataset(Dataset):
    def __init__(self, size=100000, seq_len=506):
        self.size = size
        self.seq_len = seq_len

    def __len__(self):
        return self.size

    def __getitem__(self, idx):
        # Simulate long sequences near the 512 limit
        input_ids = torch.randint(0, 20000, (self.seq_len,))
        attention_mask = torch.ones((self.seq_len,))

        # Simulate numerical instability (intermittent Inf/NaN)
        # This sometimes triggers illegal memory access in C++ extensions/kernels
        labels = torch.tensor([float(idx % 10)])
        if idx % 500 == 0:
            labels = torch.tensor([float("inf")])

        return {
            "input_ids": input_ids,
            "attention_mask": attention_mask,
            "labels": labels,
        }


def collate_fn(batch):
    # Standard collation but with large tensors
    return {key: torch.stack([item[key] for item in batch]) for key in batch[0].keys()}


if __name__ == "__main__":
    dataset = StressDataset()
    # Using high num_workers and pin_memory to maximize shared memory usage
    dataloader = DataLoader(
        dataset, batch_size=64, num_workers=4, pin_memory=True, collate_fn=collate_fn
    )

    print(f"Starting Stress Test (PID: {os.getpid()})...")
    try:
        for epoch in range(10):
            for i, batch in enumerate(dataloader):
                if i % 100 == 0:
                    # Simulate a tiny bit of GPU work
                    _ = batch["input_ids"].to("cuda")
                    if i % 1000 == 0:
                        print(f"Epoch {epoch}, Step {i} processed successfully.")
    except Exception as e:
        print(f"Caught Python Exception: {e}")
    # Note: If it's a Segfault, Python won't catch it; the process will just die.
```

  3. Steps to Reproduce
   1. Run the script on a machine with kernel.split_lock_mitigate=1.
   2. Observe the output. The crash is intermittent (non-deterministic).
   3. Monitor system logs simultaneously in another terminal:

   1     # Look for "split_lock" or "segfault"
   2     sudo journalctl -kf | grep -iE "split lock|segfault"

  4. Expected vs. Actual Behavior
   * Expected: The DataLoader should either raise a Python ValueError (for Inf values if handled) or complete the loop.
   * Actual: The worker process terminates abruptly with ERROR: Unexpected segmentation fault encountered in worker or DataLoader worker exited unexpectedly. No Python Traceback is
     provided because the crash happens in the C++/CUDA layer or is induced by a kernel-level SIGKILL/SIGBUS due to split-lock detection.

  ---

  Why these steps? 
   * seq_len=506: Long sequences increase the probability of unaligned memory access in vectorized operations.
   * split_lock_mitigate=1: On modern Intel CPUs, atomic operations crossing cache lines (split locks) are penalized. If PyTorch's internal buffer management or a specific CUDA kernel
     performs an unaligned atomic op, the kernel mitigation might terminate the process or cause enough delay to trigger a race condition in the multiprocessing queue.
   * No Traceback: The lack of a traceback is a hallmark of a signal-level crash (SIGSEGV) rather than a logic-level Python exception.
